src/NHibernate/Multi/MultiAnyQueryBase.cs

```diff
@@ -0,0 +1,217 @@
+			{
+				var qi = _queryInfos[index];
+				var resultsFromCache = qi.Loader.GetResultsIfCacheable(Session, qi.Parameters, out IQueryCache cache, out QueryKey key, qi.QuerySpaces, null);
+				qi.CacheTransformer = key?.ResultTransformer;
+
+				if (resultsFromCache != null)
+				{
+					_loaderResults[index] = resultsFromCache;
+					continue;
+				}
+
+				if (cache != null)
+				{
+					qi.PutInCacheAction = (list) => qi.Loader.PutResultInQueryCache(Session, qi.Parameters, cache, key, list);
+				}
+
+				yield return qi.Loader.CreateSqlCommand(qi.Parameters, Session);
+			}
+		}
+
+		public IEnumerable<Func<DbDataReader, int>> GetProcessResultSetActions()
+		{
+			var dialect = Session.Factory.Dialect;
+
+			for (var i = 0; i < _queryInfos.Count; i++)
+			{
+				Loader.Loader loader = _queryInfos[i].Loader;
+				var queryParameters = _queryInfos[i].Parameters;
+
+				//Skip processing for items already loaded from cache
+				if (_queryInfos[i].CacheTransformer != null && _loaderResults[i] != null)
+				{
+					loader.ProcessCachedResults(queryParameters, _queryInfos[i].CacheTransformer, ref _loaderResults[i]);
+					continue;
+				}
+
+				int entitySpan = loader.EntityPersisters.Length;
+				_hydratedObjects[i] = entitySpan == 0 ? null : new List<object>(entitySpan);
+				EntityKey[] keys = new EntityKey[entitySpan];
+
+				RowSelection selection = queryParameters.RowSelection;
+				bool createSubselects = loader.IsSubselectLoadingEnabled;
+
+				_subselectResultKeys[i] = createSubselects ? new List<EntityKey[]>() : null;
+				int maxRows = Loader.Loader.HasMaxRows(selection) ? selection.MaxRows : int.MaxValue;
+				bool advanceSelection = !dialect.SupportsLimitOffset || !loader.UseLimit(selection, dialect);
+
+				var tmpResults = new List<object>();
+				var index = i;
+				yield return reader =>
+				{
+					_reader = reader;
+					if (advanceSelection)
+					{
+						Loader.Loader.Advance(reader, selection);
+					}
+					if (queryParameters.HasAutoDiscoverScalarTypes)
+					{
+						loader.AutoDiscoverTypes(reader, queryParameters, null);
+					}
+
+					LockMode[] lockModeArray = loader.GetLockModes(queryParameters.LockModes);
+					EntityKey optionalObjectKey = Loader.Loader.GetOptionalObjectKey(queryParameters, Session);
+					int rowCount = 0;
+
+					int count;
+					for (count = 0; count < maxRows && reader.Read(); count++)
+					{
+						rowCount++;
+
+						object o =
+							loader.GetRowFromResultSet(
+								reader,
+								Session,
+								queryParameters,
+								lockModeArray,
+								optionalObjectKey,
+								_hydratedObjects[index],
+								keys,
+								true,
+								_queryInfos[index].CacheTransformer
+							);
+						if (loader.IsSubselectLoadingEnabled)
+						{
+							_subselectResultKeys[index].Add(keys);
+							keys = new EntityKey[entitySpan]; //can't reuse in this case
+						}
+
+						tmpResults.Add(o);
+					}
+					return rowCount;
+				};
+
+				_loaderResults[index] = tmpResults;
```

Comment:
Putting this after the yield leaves the state a bit inconsistent between each `MoveNext` when enumerating. Better move that before the yield.